### ISSUE TITLE
question

### ISSUE THREAD

**Issue by nry123** [milesgranger/gap_statistic#53]
# Why the results are different every time the program runs based on the same data and OptimalK object. Even using the examples you provide, the results are also different.
# How can I get the same/definite result? 
# Moreover, How do I use "clusterer_kwargs", Would you like show me an example?

**Comment by milesgranger** [comment]
> Why the results are different every time the program runs based on the same data and OptimalK object. Even using the examples you provide, the results are also different.

I don't know, this hasn't been my experience. If the dataset isn't 'clustered' then you'll likely experience variations in the cluster count due to random initialization and convergence possibly.

> How can I get the same/definite result?

If I use a seed in generating blobs, I get the same result every time... so again, maybe point one is the issue?

> Moreover, How do I use "clusterer_kwargs", Would you like show me an example?

Please look at [the example notebook](https://github.com/milesgranger/gap_statistic/blob/master/Example.ipynb), you'll see the last example shows using your own clustering function. `clusterer_kwargs` is simply any additional kwargs that your function might need during iterations.

**Comment by nry123** [comment]
I try again, but the result is also different. Please give help.


%matplotlib inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from gap_statistic import OptimalK
try:
    from sklearn.datasets.samples_generator import make_blobs
except ImportError:
    from sklearn.datasets import make_blobs
from sklearn.cluster import KMeans

#%%

optimalK = OptimalK(parallel_backend='rust')
optimalK

#%%

X, y = make_blobs(n_samples=int(1e5), n_features=2, centers=3, random_state=25)
print('Data shape: ', X.shape)

#%%

# copy data for comparision

#%%

XX=X.copy()
yy=y.copy()

#%%

# test the first exmaple 

#%%

## The first run about gap statistic
n_clusters = optimalK(X, cluster_array=np.arange(1, 15))
print('Optimal clusters: ', n_clusters)

plt.plot(optimalK.gap_df.n_clusters, optimalK.gap_df.gap_value, linewidth=3)
plt.scatter(optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].n_clusters,
            optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].gap_value, s=250, c='r')
plt.grid(True)
plt.xlabel('Cluster Count')
plt.ylabel('Gap Value')
plt.title('Gap Values by Cluster Count')
plt.show()

#%%

## The first run about gap statistic
(XX==X).all()
(yy==y).all()

#%%

## The first run about gap statistic
n_clusters = optimalK(X, cluster_array=np.arange(1, 15))
print('Optimal clusters: ', n_clusters)

plt.plot(optimalK.gap_df.n_clusters, optimalK.gap_df.gap_value, linewidth=3)
plt.scatter(optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].n_clusters,
            optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].gap_value, s=250, c='r')
plt.grid(True)
plt.xlabel('Cluster Count')
plt.ylabel('Gap Value')
plt.title('Gap Values by Cluster Count')
plt.show()

#%%

## verify the data is not changed. 
(XX==X).all()
(yy==y).all()

#%%

## The first run about gap statistic
n_clusters = optimalK(X, cluster_array=np.arange(1, 15))
print('Optimal clusters: ', n_clusters)

plt.plot(optimalK.gap_df.n_clusters, optimalK.gap_df.gap_value, linewidth=3)
plt.scatter(optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].n_clusters,
            optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].gap_value, s=250, c='r')
plt.grid(True)
plt.xlabel('Cluster Count')
plt.ylabel('Gap Value')
plt.title('Gap Values by Cluster Count')
plt.show()

#%%

# test the second exmaple, I will test another exmple by fixing random_state and defining the OptimalK instance and passing in our own clustering function

#%%

def special_clustering_func(X, k):
    m = KMeans(random_state=0)
    m.fit(X)
    return m.cluster_centers_, m.predict(X)

optimalk = OptimalK(clusterer=special_clustering_func)

#%%

## verify the data is not changed. 
(XX==X).all()
(yy==y).all()

#%%

## The first run about gap statistic
n_clusters = optimalk(X, n_refs=3, cluster_array=range(1, 4))

plt.plot(optimalK.gap_df.n_clusters, optimalK.gap_df.gap_value, linewidth=3)
plt.scatter(optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].n_clusters,
            optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].gap_value, s=250, c='r')
plt.grid(True)
plt.xlabel('Cluster Count')
plt.ylabel('Gap Value')
plt.title('Gap Values by Cluster Count')
plt.show()

#%%

## verify the data is not changed. 
(XX==X).all()
(yy==y).all()

#%%

## The second run about gap statistic
n_clusters = optimalk(X, n_refs=3, cluster_array=range(1, 4))

plt.plot(optimalK.gap_df.n_clusters, optimalK.gap_df.gap_value, linewidth=3)
plt.scatter(optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].n_clusters,
            optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].gap_value, s=250, c='r')
plt.grid(True)
plt.xlabel('Cluster Count')
plt.ylabel('Gap Value')
plt.title('Gap Values by Cluster Count')
plt.show()

#%%

## verify the data is not changed. 
(XX==X).all()
(yy==y).all()

#%%

## The third run about gap statistic
n_clusters = optimalk(X, n_refs=3, cluster_array=range(1, 4))

plt.plot(optimalK.gap_df.n_clusters, optimalK.gap_df.gap_value, linewidth=3)
plt.scatter(optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].n_clusters,
            optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].gap_value, s=250, c='r')
plt.grid(True)
plt.xlabel('Cluster Count')
plt.ylabel('Gap Value')
plt.title('Gap Values by Cluster Count')
plt.show()

#%%
the three picture is derived in the first example in orderly.
![image](https://user-images.githubusercontent.com/99719999/154877004-1fc3b345-081a-49a7-b79c-01847f4c9d20.png)
![image](https://user-images.githubusercontent.com/99719999/154877027-fff844c3-85f7-432d-bf99-1238186dbbbb.png)
![image](https://user-images.githubusercontent.com/99719999/154877048-c1e5cb82-b4c2-46c2-9e30-7ec1dd444a25.png)
the next three picture is derived in the second example in orderly.
![image](https://user-images.githubusercontent.com/99719999/154877071-93dba389-793a-4437-b2c3-82b3b81fd067.png)
![image](https://user-images.githubusercontent.com/99719999/154877086-1739feb3-bf4b-4869-b41e-9aed962e0a01.png)
![image](https://user-images.githubusercontent.com/99719999/154877102-36d39d24-078c-49e9-bc0c-8986916e4bdd.png)

**Comment by nry123** [comment]
Sorry! The last post information is wrong, please look at the content of this post.
I try again, but the result is also different. Please give me helping. Thank much.

%matplotlib inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from gap_statistic import OptimalK
try:
    from sklearn.datasets.samples_generator import make_blobs
except ImportError:
    from sklearn.datasets import make_blobs
from sklearn.cluster import KMeans

#%%

optimalK = OptimalK(parallel_backend='rust')
optimalK

#%%

X, y = make_blobs(n_samples=int(1e5), n_features=2, centers=3, random_state=25)
print('Data shape: ', X.shape)

#%%

# copy data for comparision
XX=X.copy()
yy=y.copy()

#%%

# test the first exmaple 

#%%

## The first run about gap statistic
n_clusters = optimalK(X, cluster_array=np.arange(1, 15))
print('Optimal clusters: ', n_clusters)

plt.plot(optimalK.gap_df.n_clusters, optimalK.gap_df.gap_value, linewidth=3)
plt.scatter(optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].n_clusters,
            optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].gap_value, s=250, c='r')
plt.grid(True)
plt.xlabel('Cluster Count')
plt.ylabel('Gap Value')
plt.title('Gap Values by Cluster Count')
plt.show()

#%%

## The first run about gap statistic
(XX==X).all()
(yy==y).all()

#%%

## The second run about gap statistic
n_clusters = optimalK(X, cluster_array=np.arange(1, 15))
print('Optimal clusters: ', n_clusters)

plt.plot(optimalK.gap_df.n_clusters, optimalK.gap_df.gap_value, linewidth=3)
plt.scatter(optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].n_clusters,
            optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].gap_value, s=250, c='r')
plt.grid(True)
plt.xlabel('Cluster Count')
plt.ylabel('Gap Value')
plt.title('Gap Values by Cluster Count')
plt.show()

#%%

## verify the data is not changed. 
(XX==X).all()
(yy==y).all()

#%%

## The third run about gap statistic
n_clusters = optimalK(X, cluster_array=np.arange(1, 15))
print('Optimal clusters: ', n_clusters)

plt.plot(optimalK.gap_df.n_clusters, optimalK.gap_df.gap_value, linewidth=3)
plt.scatter(optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].n_clusters,
            optimalK.gap_df[optimalK.gap_df.n_clusters == n_clusters].gap_value, s=250, c='r')
plt.grid(True)
plt.xlabel('Cluster Count')
plt.ylabel('Gap Value')
plt.title('Gap Values by Cluster Count')
plt.show()

the three picture is derived in the first example in orderly.
![image](https://user-images.githubusercontent.com/99719999/154878681-cf3b3630-eab7-446d-812d-76e9e9934967.png)
![image](https://user-images.githubusercontent.com/99719999/154878688-c69411e6-8d11-43aa-8a15-26641f7c6c26.png)
![image](https://user-images.githubusercontent.com/99719999/154878705-774eee4d-2f41-4ec7-904f-5ca74b97cf4c.png)


#%%

# test the second exmaple, I will test another exmple by fixing random_state and defining the OptimalK instance and passing in our own clustering function

#%%

def special_clustering_func(X, k):
    m = KMeans(random_state=0)
    m.fit(X)
    return m.cluster_centers_, m.predict(X)

optimalk = OptimalK(clusterer=special_clustering_func)

#%%

## verify the data is not changed. 
(XX==X).all()
(yy==y).all()

#%%

## The first run about gap statistic
n_clusters = optimalk(X, cluster_array=range(1, 15))

plt.plot(optimalk.gap_df.n_clusters, optimalk.gap_df.gap_value, linewidth=3)
plt.scatter(optimalk.gap_df[optimalk.gap_df.n_clusters == n_clusters].n_clusters,
            optimalk.gap_df[optimalk.gap_df.n_clusters == n_clusters].gap_value, s=250, c='r')
plt.grid(True)
plt.xlabel('Cluster Count')
plt.ylabel('Gap Value')
plt.title('Gap Values by Cluster Count')
plt.show()

#%%

## verify the data is not changed. 
(XX==X).all()
(yy==y).all()

#%%

## The second run about gap statistic
n_clusters = optimalk(X, cluster_array=range(1, 15))

plt.plot(optimalk.gap_df.n_clusters, optimalk.gap_df.gap_value, linewidth=3)
plt.scatter(optimalk.gap_df[optimalk.gap_df.n_clusters == n_clusters].n_clusters,
            optimalk.gap_df[optimalk.gap_df.n_clusters == n_clusters].gap_value, s=250, c='r')
plt.grid(True)
plt.xlabel('Cluster Count')
plt.ylabel('Gap Value')
plt.title('Gap Values by Cluster Count')
plt.show()

#%%

## verify the data is not changed. 
(XX==X).all()
(yy==y).all()

#%%

## The third run about gap statistic
n_clusters = optimalk(X, cluster_array=range(1, 15))

plt.plot(optimalk.gap_df.n_clusters, optimalk.gap_df.gap_value, linewidth=3)
plt.scatter(optimalk.gap_df[optimalk.gap_df.n_clusters == n_clusters].n_clusters,
            optimalk.gap_df[optimalk.gap_df.n_clusters == n_clusters].gap_value, s=250, c='r')
plt.grid(True)
plt.xlabel('Cluster Count')
plt.ylabel('Gap Value')
plt.title('Gap Values by Cluster Count')
plt.show()

#%%
the three picture is derived in the second example in orderly.
![image](https://user-images.githubusercontent.com/99719999/154878741-234d234f-ac5b-4547-bd6a-481eedc036e0.png)
![image](https://user-images.githubusercontent.com/99719999/154878763-0f8d89d6-9bf5-4223-8503-770b2cd1c1f1.png)
![image](https://user-images.githubusercontent.com/99719999/154878778-b4905e3e-71d8-4b8c-b35b-d1ebbdda8434.png)
